### PR TITLE
feat(dashboard): add jsonb filters to database page

### LINKED ISSUE #3982: dashboard: add jsonb filters, same as permissions

### PR DESCRIPTION
**Pull request by dbm03**
Resolves #3982

### Checklist

- [X] No breaking changes
- [X] Tests pass
- [X] New features have new tests
- [X] Documentation is updated (if applicable)
- [X] Title of the PR is in the correct format (see below)

### **PR Type**
Enhancement

___

### **Description**
- Add JSONB-specific filter operators (`@>`, `<@`, `?`, `?|`, `?&`) to the database data browser, enabling PostgreSQL-native containment and key-existence filtering on JSONB columns.
- Make the operator dropdown column-type-aware: JSONB columns show only common + JSONB operators, while other columns show common + comparison + text operators. Switching to a column whose type is incompatible with the current operator automatically resets the operator and value.
- Generate correct SQL WHERE clauses for JSONB operators, including `::jsonb` casts for `@>`/`<@` and `array[...]` formatting for `?|`/`?&`.
- Extract `useGetDataColumns` hook to deduplicate the "all columns minus selection column" logic used across filter components.
- Add `validOperators` set to filter stale/invalid operators from localStorage, preventing crashes when stored filters reference removed operators.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["DataGridFilter"] -->|"columnDataType"| B["DataGridFilterOperators"]
  A -->|"currentOp"| C["DataGridFilterColumn"]
  C -->|"on column change"| D["getAvailableOperators()"]
  D -->|"resets op if incompatible"| C
  B -->|"renders ops for type"| D
  E["filtersToWhere()"] -->|"@>/<@ → ::jsonb cast\n?|/?& → array[] format"| F["SQL WHERE clause"]
  G["PersistentDataGridFilterStorage"] -->|"validates via validOperators"| E
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody>
<tr><td><strong>Core logic</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>operators.ts</strong><dd><code>Reorganize operators into category arrays (common, comparison, text, jsonb); add getAvailableOperators() and validOperators set</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-13294a141673a104dcd136b2e8f9b6dacd2f2ee267382d342c67b71cc5b10cee">+46/-3</a></td>
</tr>
<tr>
  <td><strong>filtersToWhere.ts</strong><dd><code>Add SQL generation for JSONB operators: ::jsonb cast for @>/<@ and array[] formatting for ?|/?&</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-f70c193a85607e6240cf4c9cfb7b77f947d45ba0f1f2759a277ea3c2bf11c95f">+7/-0</a></td>
</tr>
<tr>
  <td><strong>PersistentDataGridFilterStorage.ts</strong><dd><code>Validate stored filters against validOperators to drop stale entries on load</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-a759a39c36ba2bac7ee37bfa344b66221a6a3ae4a25e301c6ec36f516c28d1dd">+4/-1</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>UI wiring</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>DataGridFilter.tsx</strong><dd><code>Resolve selected column's dataType and pass it to operator/column subcomponents</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-fe80725c4a49e01acea976157349af6544419196bff458b5b013afd27bef30e0">+10/-2</a></td>
</tr>
<tr>
  <td><strong>DataGridFilterColumn.tsx</strong><dd><code>Reset operator and value when switching to a column whose type doesn't support the current operator</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-33237a95dff61892aa847c1850f5766d16da8d75751d0a16284684f9d8d56cb2">+20/-9</a></td>
</tr>
<tr>
  <td><strong>DataGridFilterOperators.tsx</strong><dd><code>Accept columnDataType prop; use getAvailableOperators() to render type-appropriate operators; set default values for @>/<@</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-2212b494ff381e5aac30a2261f6095e32a32836a4bb308afe5409e7018c9fdd1">+8/-3</a></td>
</tr>
<tr>
  <td><strong>DataGridFilterActions.tsx</strong><dd><code>Replace inline column filtering with useGetDataColumns hook</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-a757ce03ed5f1747eb8765cc195e0e20a4881895355e27278e9bab1c2afab398">+2/-6</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Shared hooks</strong></td><td><details><summary>1 file</summary><table>
<tr>
  <td><strong>useGetDataColumns.ts</strong><dd><code>New hook extracting "all columns minus selection column" logic shared across filter components</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-48a5c0cd5ffa47d053be3a5b8834cbebef335528dfc40097a70c1d3c618a07bf">+14/-0</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>operators.test.ts</strong><dd><code>Tests for getAvailableOperators: verifies correct operator sets for jsonb, non-jsonb, and undefined data types</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-0fcdfc615a82c0e435edc6e49e56939ac7e3f7afd491c4274ab809a5aec37061">+61/-0</a></td>
</tr>
<tr>
  <td><strong>filtersToWhere.test.ts</strong><dd><code>Tests for JSONB WHERE clause generation: @>, <@, ?, ?|, ?& operators including edge cases (empty object, spaces around commas)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3998/files#diff-d651d7d579403c7b8e75b85552c6cbf5f16783d5ea732461ec5f4279f2ecd8d8">+86/-0</a></td>
</tr>
</table></details></td></tr>
</tbody></table>

</details>

___

### CODE REVIEW

File: dashboard/src/features/orgs/projects/common/components/DataGridFiltersPopover/DataGridFilter.tsx [hunk 1]
```diff
@@ -15,11 +17,21 @@ type FilterProps = {
 
 function DataGridFilter({ column, op, value, index }: FilterProps) {
   const { removeFilter } = useDataGridFilters();
+  const { getAllColumns } = useDataGridConfig<{ dataType: string }>();
+  const columns = getAllColumns().filter(
```
Comment (robertkasza) [inline dashboard/src/features/orgs/projects/common/components/DataGridFiltersPopover/DataGridFilter.tsx]: I believe this logic is used in a couple of places, what do you think about moving this logic into a hook an used that here?

Something like this:
```
function useColumns<T>() {
  const { getAllColumns } = useDataGridConfig<T>();
  const columns = getAllColumns().filter(
    ({ id }) => id !== SELECTION_COLUMN_ID,
  );
  
  return columns;
}
```

useColumns might be not the best name or we use it for something else.

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts [hunk 2]
```diff
@@ -21,4 +27,33 @@ export const operators = [
...
+  ...jsonbOperators,
+] as const;
+
 export type DataGridFilterOperator = (typeof operators)[number]['op'];
+
+export const validOperators: Set<string> = new Set(
+  operators.map((operator) => operator.op),
+);
+
+export function getAvailableOperators(
```
Comment (robertkasza) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts:59]: Whenever a component is rerendered a new array is created which can cause unnecessary renders. Let's use the same logic like in here: 

https://github.com/nhost/nhost/blob/main/dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/getAvailableOperators.ts#L63

Create constants for the jsonb operators and the other and return those that way the array will not change.

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts [hunk 3]
```diff
@@ -21,4 +27,33 @@ export const operators = [
   { op: '!~*', label: 'not regex (case-insensitive)' },
 ] as const;
 
+const jsonbOperators = [
+  { op: '@>', label: 'contains (jsonb)' },
```
Comment (robertkasza) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts]: Do we really need the `(jsonb)` text in the label? It seems a little bit redundant to me, but If you think it gives valuable information to the user, then I am fine leaving it.

File: dashboard/src/features/orgs/projects/database/dataGrid/hooks/useTableQuery/filtersToWhere.ts [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
+import { validOperators } from '@/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider';
 import { isEmptyValue } from '@/lib/utils';
 
 export function filtersToWhere(filters?: DataGridFilter[]): string {
   if (isEmptyValue(filters)) {
     return '';
   }
 
-  const whereClauses = filters!.map((filter) => {
+  const validFilters = filters!.filter((filter) =>
```
Comment (robertkasza) [inline dashboard/src/features/orgs/projects/database/dataGrid/hooks/useTableQuery/filtersToWhere.ts]: I think this is not right. At this point we should not need to validate anything, just convert the filters to string.

I am not even sure if we can ever have an invalid operator, can we?

File: dashboard/src/features/orgs/projects/common/components/DataGridFiltersPopover/DataGridFilter.tsx (hunk 1 above)
Comment (dbm03) [reply dashboard/src/features/orgs/projects/common/components/DataGridFiltersPopover/DataGridFilter.tsx]: I'm changing it in:

4d06e08

I've named it `useGetDataColumns`, because it gets all the columns that have actual data, not the selection column with the checkbox.

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts (hunk 2 above)
Comment (dbm03) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts:59]: I'm doing this now in b74c30d:

```ts
const jsonbColumnOperators = [...commonOperators, ...jsonbOperators] as const;

const defaultColumnOperators = [
  ...commonOperators,
  ...comparisonOperators,
  ...textOperators,
] as const;

export function getAvailableOperators(
  dataType?: string,
): readonly { op: DataGridFilterOperator; label: string }[] {
  if (dataType === 'jsonb') {
    return jsonbColumnOperators;
  }

  return defaultColumnOperators;
}
```

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts (hunk 3 above)
Comment (dbm03) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts]: I think I can probably just remove the `(jsonb)` text and it's still understandable.
image

Do you think adding a group label makes it any better? If so I can add it

e.g.

image

File: dashboard/src/features/orgs/projects/database/dataGrid/hooks/useTableQuery/filtersToWhere.ts (hunk 4 above)
Comment (dbm03) [reply dashboard/src/features/orgs/projects/database/dataGrid/hooks/useTableQuery/filtersToWhere.ts]: My concern was, what if for some reason a user has a manipulated localStorage with an invalid operator: it could be any SQL string that does anything in the database.

But yeah, it's true we should assume that props for `filtersToWhere` are fine/validated.

I'm going to add this instead to the PersistentDataGridFilterStorage file, so we validate operators exactly when filters are parsed from localStorage
```ts
export function getDataGridFilters(tablePath: string): DataGridFilter[] {
  const allStoredData = getAllStoredData();
  const filters = allStoredData[tablePath] ?? [];

  return filters.filter((filter) => validOperators.has(filter.op));
}
```

8ee42d1

File: dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts (hunk 3 above)
Comment (robertkasza) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/DataBrowserGrid/DataGridQueryParamsProvider/operators.ts]: I am not sure, i would leave it as is now. (No grouping)
